components/callbacks/nexus_invocation.go

```diff
@@ -0,0 +1,132 @@
+// The MIT License
+//
+// Copyright (c) 2024 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package callbacks
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"slices"
+	"time"
+
+	"github.com/nexus-rpc/sdk-go/nexus"
+
+	persistencepb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/service/history/queues"
+)
+
+var retryable4xxErrorTypes = []int{
+	http.StatusRequestTimeout,
+	http.StatusTooManyRequests,
+}
+
+type CanGetNexusCompletion interface {
+	GetNexusCompletion(ctx context.Context) (nexus.OperationCompletion, error)
+}
+
+type nexusInvocation struct {
+	nexus      *persistencepb.Callback_Nexus
+	completion nexus.OperationCompletion
+}
+
+func isRetryableHTTPResponse(response *http.Response) bool {
+	return response.StatusCode >= 500 || slices.Contains(retryable4xxErrorTypes, response.StatusCode)
+}
+
+func outcomeTag(callCtx context.Context, response *http.Response, callErr error) string {
+	if callErr != nil {
+		if callCtx.Err() != nil {
+			return "request-timeout"
+		}
+		return "unknown-error"
+	}
+	return fmt.Sprintf("status:%d", response.StatusCode)
+}
+
+func (n nexusInvocation) WrapError(result invocationResult, err error) error {
+	// If the request permanently failed there is no need to raise the error
+	if result == failed {
+		return nil
+	}
+	if err != nil {
+		return queues.NewDestinationDownError(err.Error(), err)
```

Comment:
We previously had this: `queues.NewDestinationDownError(callErr.Error(), err)`, now the message is from 
the `saveResult` error.

Reply:
https://github.com/tli-temporal/temporal/blob/hsm-callback/components/callbacks/executors.go#L116

This err is from the invocation.